sample-tracking-status-overview-domain/src/test/groovy/life/qbic/business/project/subscribe/SubscribeProjectSpec.groovy

```diff
@@ -0,0 +1,99 @@
+package life.qbic.business.project.subscribe
+
+import life.qbic.business.DataSourceException
+import spock.lang.Specification
+
+/**
+ * <b>Specification for the SubscribeProject use case</b>
+ *
+ * @since 1.0.0
+ */
+class SubscribeProjectSpec extends Specification {
+    final String validFirstName = "firstname"
+    final String validLastName = "lastname"
+    final String validEmail = "email@addre.ss"
+    final String validProjectCode = "QABCD"
+
+    SubscribeProjectOutput output = Mock()
+    SubscriptionDataSource subscriptionDataSource = Mock()
+    SubscribeProject subscribeProject = new SubscribeProject(subscriptionDataSource, output)
+
+    def "Subscribe fails for invalid first name: #invalidFirstName"() {
+        when:
+        subscribeProject.subscribe(invalidFirstName, validLastName, validEmail, validProjectCode)
+        then:
+        thrown(IllegalArgumentException)
+        where:
+        invalidFirstName << [null, ""]
+    }
+
+    def "Subscribe fails for invalid last name: #invalidLastName"() {
+        when:
+        subscribeProject.subscribe(validFirstName, invalidLastName, validEmail, validProjectCode)
+        then:
+        thrown(IllegalArgumentException)
+        where:
+        invalidLastName << [null, ""]
+    }
+
+    def "Subscribe fails for invalid email address: #invalidEmail"() {
+        when:
+        subscribeProject.subscribe(validFirstName, validLastName, invalidEmail, validProjectCode)
+        then:
+        thrown(IllegalArgumentException)
+        where:
+        invalidEmail << [null, ""]
+    }
+
+    def "Subscribe fails for invalid project code: #invalidProjectCode"() {
+        when:
+        subscribeProject.subscribe(validFirstName, validLastName, validEmail, invalidProjectCode)
+        then:
+        thrown(IllegalArgumentException)
+        where:
+        invalidProjectCode << [null, "", "1234", "ZBCA"]
+    }
+
+    def "Subscribe does not throw an IllegalArgumentException for valid arguments"() {
+        when:
+        subscribeProject.subscribe(validFirstName, validLastName, validEmail, validProjectCode)
+        then:
+        notThrown(IllegalArgumentException)
+    }
```

Comment:
No, we need to assert that our valid arguments are perceived as valid. 
This test covers all possibilities where the validation of the input parameters throws an error but there is no error. We need to know that we can rely on the input parameters being validated correctly.

Reply:
I do not think that this will lead to major obfuscation in the tests since those small tests are pretty obvious.

Reply:
No it does not lead to confusion but I was wondering if we really need it. And afaik it only tests if no error is thrown for the provided arguments.. which are the same arguments you use in the test below. We check if the correct output method is called, so I see it as coverd. Nevertheless, I will not block you PR because of this. I just wanted to make you aware of it.